base/strings/string.jl

```diff
@@ -300,29 +402,90 @@ end
+#  currently it uses a fast loop to scan for non ascii characters then when it encounters a
+#  multibyte character it process only a single multbyte character before going back to look
+#  for non-ascii characters.  A more balanced algorithm would likely want to process multibyte
+#  characters in blocks of 64 bytes
+function _length_nonascii_decrement(
+    cu::AbstractVector{UInt8}, first::Int, last::Int, c::Int, state=_GUTF8_DFA_ACCEPT
+)
+    state = ifelse(state == _GUTF8_DFA_INVALID, _GUTF8_DFA_ACCEPT, state)
+    i = ifelse(state == _GUTF8_DFA_ACCEPT, first - 1, first)
+    #@inbounds b = codeunit(s, first)
+    @inbounds b = cu[first]
+    @inbounds while true
+        #This logic enables the first state to be >_GUTF8_DFA_INVALID so that a chunk
+        # can continue from a previous chunk
+        (state == _GUTF8_DFA_ACCEPT) && (i += 1)
+        #Logic was taken out of the n=1:3 loop below so we must correct the count here
+        (state == _GUTF8_DFA_INVALID) && (c += 1)
+        if state <= _GUTF8_DFA_INVALID
+            #Loop through all the one byte characters
+            while true
+                b = cu[i]
+                ((i += 1) <= last) || break
+                0xc0 ≤ b ≤ 0xf7 && break
+            end
+            state = _gutf8_dfa_step(_GUTF8_DFA_ACCEPT, b)
+            (i <= last) || return (c, state)
+        end
+
+        #This should get unrolled
+        for n in 1:3
+            state = _gutf8_dfa_step(state, cu[i])
+            c -= 1
+            state <= _GUTF8_DFA_INVALID && break
+            ((i += 1) <= last) || return (c, state)
         end
-        l = b
-        b = codeunit(s, i) # cont byte 1
-        c -= (x = b & 0xc0 == 0x80)
-        x & (l ≥ 0xe0) || continue
-
-        (i += 1) ≤ n || return c
-        b = codeunit(s, i) # cont byte 2
-        c -= (x = b & 0xc0 == 0x80)
-        x & (l ≥ 0xf0) || continue
-
-        (i += 1) ≤ n || return c
-        b = codeunit(s, i) # cont byte 3
-        c -= (b & 0xc0 == 0x80)
     end
+    return (c, state)
+end
+
+function _length_continued_nonascii(
+    cu::AbstractVector{UInt8}, first::Int, last::Int, c::Int
+)
+    chunk_size = _STRING_LENGTH_CHUNKING_SIZE
+
+    start = first
+    stop = min(last, first + chunk_size - 1)
+    state = _GUTF8_DFA_ACCEPT
+
+    while start <= last
+        #First we process a non ascii chunk because we assume the barrier
+        # function sent it here for a reason
+        (c, state) = _length_nonascii_decrement(cu, start, stop, c, state)
+        start = start + chunk_size
+        stop = min(last, stop + chunk_size)
+
+        while state <= _GUTF8_DFA_INVALID
+            _isascii(cu, start, stop) || break
+            (start = start + chunk_size) <= last || break
+            stop = min(last, stop + chunk_size)
+        end
+    end
+    return c
+end
+
+@assume_effects :terminates_locally @inline @propagate_inbounds  function length_continued(s::String, first::Int, last::Int, c::Int)
```

Comment:
this will need `terminates_globally` rather than `terminates_locally` since it calls a bunch of functions with loops.

Reply:
The solution ended up being to just explicitly make it `:foldable`